src/api.rs

```diff
@@ -241,3 +242,43 @@ impl Drop for ThreadPool {
         self.registry.terminate();
     }
 }
+
+/// Stack-scoped thread-local storage for thread pools.
+pub struct ScopedTLS<'a, T: Send> {
+    _pool: &'a ThreadPool,
+    slots: Box<[RefCell<Option<T>>]>,
+}
+
+unsafe impl<'a, T: Send> Sync for ScopedTLS<'a, T> {}
+
+impl<'a, T: Send> ScopedTLS<'a, T> {
+    pub fn new(p: &'a ThreadPool) -> Self {
+        let mut v = Vec::new();
+        for _ in 0..p.registry.num_threads() {
+            v.push(RefCell::new(None));
+        }
+
+        ScopedTLS {
+            _pool: p,
+            slots: v.into_boxed_slice(),
+        }
+    }
+
+    pub fn borrow(&self) -> Ref<Option<T>> {
+        let idx = current_thread_index();
+        self.slots[idx].borrow()
+    }
+
+    pub fn borrow_mut(&self) -> RefMut<Option<T>> {
+        let idx = current_thread_index();
+        self.slots[idx].borrow_mut()
+    }
+}
+
+fn current_thread_index() -> usize {
+    unsafe {
+        let curr = WorkerThread::current();
```

Comment:
One catch is that we don't know the worker-thread is associated with the thread-pool that the `ScopedTLS` corresponds to. My inclination would be to use the machinery I introduced in https://github.com/nikomatsakis/rayon/pull/141 (which I should probably land at some point...), which basically maintains a linked list of registries (which is expected to be of length 1, really) to find the appropriate one. In that case, you could just create the `ScopedTLS` without giving a `ThreadPool` (since the common case is not to have one, but instead to use the global, default thread-pool).

But another option would be to assert in this method that the worker belongs to the given thread-pool (and perhaps modify `new` so that it takes no arguments, in which cases it uses the default registry, which would either be the global registry or -- if in a worker -- the current thread-pool).